In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/germany-used-cars-dataset-2023/data.csv


In [2]:
df=pd.read_csv('/kaggle/input/germany-used-cars-dataset-2023/data.csv',index_col='Unnamed: 0')
df

brand              model  color registration_date  year  \
0       alfa-romeo     Alfa Romeo GTV    red           10/1995  1995   
1       alfa-romeo     Alfa Romeo 164  black           02/1995  1995   
2       alfa-romeo  Alfa Romeo Spider  black           02/1995  1995   
3       alfa-romeo  Alfa Romeo Spider  black           07/1995  1995   
4       alfa-romeo     Alfa Romeo 164    red           11/1996  1996   
...            ...                ...    ...               ...   ...   
251074       volvo         Volvo XC40  white           04/2023  2023   
251075       volvo         Volvo XC90  white           03/2023  2023   
251076       volvo          Volvo V60  white           05/2023  2023   
251077       volvo         Volvo XC40  white           05/2023  2023   
251078       volvo         Volvo XC40   gold           03/2023  2023   

       price_in_euro power_kw power_ps transmission_type fuel_type  \
0               1300      148      201            Manual    Petrol   
1              24900      191      260            Manual    Petrol   
2               5900      110      150           Unknown    Petrol   
3               4900      110      150            Manual    Petrol   
4              17950      132      179            Manual    Petrol   
...              ...      ...      ...               ...       ...   
251074         57990      192      261         Automatic    Hybrid   
251075         89690      173      235         Automatic    Diesel   
251076         61521      145      197         Automatic    Diesel   
251077         57890      132      179         Automatic    Hybrid   
251078         52900      160      218         Automatic  Electric   

       fuel_consumption_l_100km fuel_consumption_g_km  mileage_in_km  \
0                 10,9 l/100 km              260 g/km       160500.0   
1                           NaN              - (g/km)       190000.0   
2                           NaN              - (g/km)       129000.0   
3                  9,5 l/100 km              225 g/km       189500.0   
4                  7,2 l/100 km              - (g/km)        96127.0   
...                         ...                   ...            ...   
251074                      NaN      43 km Reichweite         1229.0   
251075             7,6 l/100 km              202 g/km         4900.0   
251076             4,7 l/100 km              125 g/km         1531.0   
251077                      NaN      45 km Reichweite         1500.0   
251078                      NaN     438 km Reichweite           50.0   

                                        offer_description  
0                                               2.0 V6 TB  
1                                     Q4 Allrad, 3.2L GTA  
2                                           ALFA ROME 916  
3                                    2.0 16V Twin Spark L  
4                   3.0i Super V6, absoluter Topzustand !  
...                                                   ...  
251074  Plus Bright T5 Recharge Intellisafe*Surround+P...  
251075  B5 AWD Diesel Ultimate Dark 7-Sitzer Massage F...  
251076            B4 D Plus Dark 145 kW, 5-türig (Diesel)  
251077        T5 Recharge Plus Dark *Standh*360°*beh.Lenk  
251078               Ultimate Recharge Twin Motor AHK GJR  

[251079 rows x 14 columns]

Categorical Columns:<br/>
brand: The brand or manufacturer of the car.<br/>
model: The specific model of the car.<br/>
color: The color of the car's exterior.<br/>
registration_date: The date when the car was registered (Month/Year).<br/>
transmission_type: The type of transmission (e.g., automatic, manual).<br/>
fuel_type: The type of fuel the car requires.<br/>
offer_description: Additional description provided in the car offer.<br/>

Numerical Columns:<br/>
year: The year in which the car was manufactured.<br/>
power_kw: The power of the car in kilowatts (kW).<br/>
power_ps: The power of the car in horsepower (ps).<br/>
fuel_consumption_l_100km: Information about the car's fuel consumption in L/100km.<br/>
fuel_consumption_g_km: Information about the car's fuel consumption in g/km.<br/>
mileage_in_km: The total distance traveled by the car in km.<br/>
<b> price_in_euro</b>: The price of the car in Euros.This is the Output column<br/> 

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 251079 entries, 0 to 251078
Data columns (total 14 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   brand                     251079 non-null  object 
 1   model                     251079 non-null  object 
 2   color                     250913 non-null  object 
 3   registration_date         251075 non-null  object 
 4   year                      251079 non-null  object 
 5   price_in_euro             251079 non-null  object 
 6   power_kw                  250945 non-null  object 
 7   power_ps                  250950 non-null  object 
 8   transmission_type         251079 non-null  object 
 9   fuel_type                 251079 non-null  object 
 10  fuel_consumption_l_100km  224206 non-null  object 
 11  fuel_consumption_g_km     251079 non-null  object 
 12  mileage_in_km             250927 non-null  float64
 13  offer_description         251078 non-null  o

In [4]:
df.isnull().sum()

brand                           0
model                           0
color                         166
registration_date               4
year                            0
price_in_euro                   0
power_kw                      134
power_ps                      129
transmission_type               0
fuel_type                       0
fuel_consumption_l_100km    26873
fuel_consumption_g_km           0
mileage_in_km                 152
offer_description               1
dtype: int64

In [5]:
#To show the percentage of null values in each column
null_percentages = df.isnull().sum() / len(df) * 100
print(null_percentages)

brand                        0.000000
model                        0.000000
color                        0.066115
registration_date            0.001593
year                         0.000000
price_in_euro                0.000000
power_kw                     0.053370
power_ps                     0.051378
transmission_type            0.000000
fuel_type                    0.000000
fuel_consumption_l_100km    10.703006
fuel_consumption_g_km        0.000000
mileage_in_km                0.060539
offer_description            0.000398
dtype: float64


<p>It can be seen that highest number of null values can be seen in fuel_consumption_l_100km coulumn i.e 10.7 %</p>
<li>Columns brand, model, year, transmission_type, fuel_type, fuel_consumption_g_km, and offer_description: These columns have no missing values (0% null), so no action is required.</li>
<li>color, registration_date, power_kw, power_ps, and mileage_in_km: These columns have very low percentages of missing values.Best option is to drop the rows with missing values since the percentage is negligible (less than 0.1%)</li>
<li>price_in_euro: This column also has no missing values (0% null), so no action is required.</li>
<li>fuel_consumption_l_100km: This column has a relatively higher percentage of missing values (10.7%). You can consider using mean or median imputation to fill in the missing values. Another approach could be to explore if there are any patterns or correlations between this column </li>

In [6]:
#dropping less null containing columns

df_dropped=df.dropna(subset=['color','registration_date','year','power_kw','power_ps','mileage_in_km'])


In [7]:
df_dropped['fuel_consumption_l_100km']

0         10,9 l/100 km
1                   NaN
2                   NaN
3          9,5 l/100 km
4          7,2 l/100 km
              ...      
251074              NaN
251075     7,6 l/100 km
251076     4,7 l/100 km
251077              NaN
251078              NaN
Name: fuel_consumption_l_100km, Length: 250629, dtype: object

In [8]:
import pandas as pd
substring = "l/100 km"
matches = df_dropped['fuel_consumption_l_100km'].str.contains(substring,na=True)
matches.value_counts()


True     249474
False      1155
Name: fuel_consumption_l_100km, dtype: int64

This shows that 1155 contains units other than l/100 km i.e other units are dropped for example:::
<li>km (Ort)</li>
<li>g/km</li>
<li>kWh/100 km</li>


In [9]:
filtered_df = df_dropped[matches==True]


In [10]:
import pandas as pd
filtered_df.loc[:, 'fuel_consumption_l_100km'] = filtered_df['fuel_consumption_l_100km'].str.replace('l/100 km', '').str.replace(",", '.')
filtered_df.loc[:, 'fuel_consumption_l_100km'] = filtered_df['fuel_consumption_l_100km'].replace('- ()', np.nan)
filtered_df.loc[:, 'fuel_consumption_l_100km'] = filtered_df['fuel_consumption_l_100km'].str.replace('l/100 km', '').astype(float)
mean_val=filtered_df['fuel_consumption_l_100km'].dropna().mean()



/tmp/ipykernel_20/1814833613.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df.loc[:, 'fuel_consumption_l_100km'] = filtered_df['fuel_consumption_l_100km'].str.replace('l/100 km', '').str.replace(",", '.')
/tmp/ipykernel_20/1814833613.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df.loc[:, 'fuel_consumption_l_100km'] = filtered_df['fuel_consumption_l_100km'].replace('- ()', np.nan)
/tmp/ipykernel_20/1814833613.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy 

In [11]:
filtered_df['fuel_consumption_l_100km'].fillna(mean_val)

0         10.900000
1          6.088828
2          6.088828
3          9.500000
4          7.200000
            ...    
251074     6.088828
251075     7.600000
251076     4.700000
251077     6.088828
251078     6.088828
Name: fuel_consumption_l_100km, Length: 249474, dtype: float64